**MLP**

In [ ]:
import torch
import torchvision
import torchvision.transforms as transforms
import torch.nn as nn
import torch.optim as optim

# Define the MLP model
class MLP(nn.Module):
    def __init__(self):
        super(MLP, self).__init__()
        self.fc1 = nn.Linear(3 * 32 * 32, 512)
        self.fc2 = nn.Linear(512, 256)
        self.fc3 = nn.Linear(256, 10)

    def forward(self, x):
        x = x.view(-1, 3 * 32 * 32)  # Flatten the input
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        x = self.fc3(x)
        return x

# Download and prepare the CIFAR-10 dataset
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
])

trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=4,
                                          shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=4,
                                         shuffle=False, num_workers=2)

classes = ('plane', 'car', 'bird', 'cat',
           'deer', 'dog', 'frog', 'horse', 'ship', 'truck')

# Initialize the MLP model, loss function, and optimizer
model = MLP()
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.001, momentum=0.9)

# Train the model
num_epochs = 10
for epoch in range(num_epochs):
    running_loss = 0.0
    for i, data in enumerate(trainloader, 0):
        inputs, labels = data
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
        if i % 2000 == 1999:
            print(f'Epoch {epoch + 1}, Batch {i + 1}: Loss {running_loss / 2000:.3f}')
            running_loss = 0.0

print('Training finished')

# Evaluate the model on the test set
correct = 0
total = 0
test_loss = 0.0

with torch.no_grad():
    for data in testloader:
        images, labels = data
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()
        loss = criterion(outputs, labels)
        test_loss += loss.item()

accuracy = 100 * correct / total
average_test_loss = test_loss / len(testloader)

print(f'Accuracy on the test set: {accuracy:.2f}%')
print(f'Average loss on the test set: {average_test_loss:.4f}')


100%|██████████| 170498071/170498071 [00:02<00:00, 77061902.82it/s]


Extracting ./data/cifar-10-python.tar.gz to ./data
Files already downloaded and verified
Epoch 1, Batch 2000: Loss 1.909
Epoch 1, Batch 4000: Loss 1.693
Epoch 1, Batch 6000: Loss 1.634
Epoch 1, Batch 8000: Loss 1.590
Epoch 1, Batch 10000: Loss 1.565
Epoch 1, Batch 12000: Loss 1.541
Epoch 2, Batch 2000: Loss 1.438
Epoch 2, Batch 4000: Loss 1.425
Epoch 2, Batch 6000: Loss 1.456
Epoch 2, Batch 8000: Loss 1.413
Epoch 2, Batch 10000: Loss 1.407
Epoch 2, Batch 12000: Loss 1.390
Epoch 3, Batch 2000: Loss 1.281
Epoch 3, Batch 4000: Loss 1.296
Epoch 3, Batch 6000: Loss 1.316
Epoch 3, Batch 8000: Loss 1.286
Epoch 3, Batch 10000: Loss 1.323
Epoch 3, Batch 12000: Loss 1.308
Epoch 4, Batch 2000: Loss 1.186
Epoch 4, Batch 4000: Loss 1.204
Epoch 4, Batch 6000: Loss 1.226
Epoch 4, Batch 8000: Loss 1.221
Epoch 4, Batch 10000: Loss 1.221
Epoch 4, Batch 12000: Loss 1.218
Epoch 5, Batch 2000: Loss 1.114
Epoch 5, Batch 4000: Loss 1.110
Epoch 5, Batch 6000: Loss 1.139
Epoch 5, Batch 8000: Loss 1.138
Epoch 5

**CNN**

In [ ]:
import torch
import torchvision
import torchvision.transforms as transforms
import torch.nn as nn
import torch.optim as optim

# Define the CNN architecture
class CNN(nn.Module):
    def __init__(self):
        super(CNN, self).__init__()
        self.conv1 = nn.Conv2d(3, 16, 3, padding=1)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(16, 32, 3, padding=1)
        self.fc1 = nn.Linear(32 * 8 * 8, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.pool(torch.relu(self.conv1(x)))
        x = self.pool(torch.relu(self.conv2(x)))
        x = x.view(-1, 32 * 8 * 8)
        x = torch.relu(self.fc1(x))
        x = self.fc2(x)
        return x

# Download and prepare the CIFAR-10 dataset
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
])

trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=4,
                                          shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=4,
                                         shuffle=False, num_workers=2)

classes = ('plane', 'car', 'bird', 'cat',
           'deer', 'dog', 'frog', 'horse', 'ship', 'truck')

# Initialize the CNN model, loss function, and optimizer
model = CNN()
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.001, momentum=0.9)

# Train the CNN model
num_epochs = 10
for epoch in range(num_epochs):
    running_loss = 0.0
    for i, data in enumerate(trainloader, 0):
        inputs, labels = data
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
        if i % 2000 == 1999:
            print(f'Epoch {epoch + 1}, Batch {i + 1}: Loss {running_loss / 2000:.3f}')
            running_loss = 0.0

print('Training finished')

# Evaluate the CNN model on the test set
correct = 0
total = 0
test_loss = 0.0

with torch.no_grad():
    for data in testloader:
        images, labels = data
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()
        loss = criterion(outputs, labels)
        test_loss += loss.item()

accuracy_cnn = 100 * correct / total
average_test_loss_cnn = test_loss / len(testloader)

print(f'Accuracy of CNN on the test set: {accuracy_cnn:.2f}%')
print(f'Average loss of CNN on the test set: {average_test_loss_cnn:.4f}')

100%|██████████| 170498071/170498071 [00:03<00:00, 48930678.52it/s]


Extracting ./data/cifar-10-python.tar.gz to ./data
Files already downloaded and verified
Epoch 1, Batch 2000: Loss 2.014
Epoch 1, Batch 4000: Loss 1.586
Epoch 1, Batch 6000: Loss 1.448
Epoch 1, Batch 8000: Loss 1.393
Epoch 1, Batch 10000: Loss 1.307
Epoch 1, Batch 12000: Loss 1.229
Epoch 2, Batch 2000: Loss 1.129
Epoch 2, Batch 4000: Loss 1.095
Epoch 2, Batch 6000: Loss 1.078
Epoch 2, Batch 8000: Loss 1.047
Epoch 2, Batch 10000: Loss 1.026
Epoch 2, Batch 12000: Loss 1.019
Epoch 3, Batch 2000: Loss 0.889
Epoch 3, Batch 4000: Loss 0.887
Epoch 3, Batch 6000: Loss 0.918
Epoch 3, Batch 8000: Loss 0.891
Epoch 3, Batch 10000: Loss 0.877
Epoch 3, Batch 12000: Loss 0.882
Epoch 4, Batch 2000: Loss 0.745
Epoch 4, Batch 4000: Loss 0.764
Epoch 4, Batch 6000: Loss 0.764
Epoch 4, Batch 8000: Loss 0.788
Epoch 4, Batch 10000: Loss 0.768
Epoch 4, Batch 12000: Loss 0.764
Epoch 5, Batch 2000: Loss 0.635
Epoch 5, Batch 4000: Loss 0.655
Epoch 5, Batch 6000: Loss 0.667
Epoch 5, Batch 8000: Loss 0.678
Epoch 5

**Transfer Learning with VGG**

In [ ]:
import torch
import torchvision
import torchvision.transforms as transforms
import torch.nn as nn
import torch.optim as optim
import torchvision.models as models

# Check if CUDA is available and set device accordingly
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f'Using device: {device}')

# Define transformations and load CIFAR-10 dataset
transform = transforms.Compose([
    transforms.Resize((224, 224)),  # Resize images to VGG input size
    transforms.ToTensor(),
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
])

trainset = torchvision.datasets.CIFAR10(root='./data', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=64,
                                          shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=64,
                                         shuffle=False, num_workers=2)

# Load pre-trained VGG model and modify for CIFAR-10
vgg_model = models.vgg16(pretrained=True)
vgg_model.to(device)  # Move model to GPU if available

for param in vgg_model.parameters():
    param.requires_grad = False

num_features = vgg_model.classifier[6].in_features
vgg_model.classifier[6] = nn.Linear(num_features, 10)  # 10 classes in CIFAR-10
vgg_model.to(device)  # Move modified model to GPU if available

# Define loss function, optimizer, and number of epochs
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(vgg_model.classifier.parameters(), lr=0.001, momentum=0.9)
num_epochs = 10

# Train the adapted VGG model
vgg_model.train()
for epoch in range(num_epochs):
    running_loss = 0.0
    for i, data in enumerate(trainloader, 0):
        inputs, labels = data
        inputs, labels = inputs.to(device), labels.to(device)  # Move data to GPU if available
        optimizer.zero_grad()
        outputs = vgg_model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
        if i % 2000 == 1999:
            print(f'Epoch {epoch + 1}, Batch {i + 1}: Loss {running_loss / 2000:.3f}')
            running_loss = 0.0

print('Training finished')

# Evaluate the adapted VGG model on the test set
correct = 0
total = 0
test_loss = 0.0

vgg_model.eval()
with torch.no_grad():
    for data in testloader:
        images, labels = data
        images, labels = images.to(device), labels.to(device)  # Move data to GPU if available
        outputs = vgg_model(images)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()
        loss = criterion(outputs, labels)
        test_loss += loss.item()

accuracy_vgg = 100 * correct / total
average_test_loss_vgg = test_loss / len(testloader)

print(f'Accuracy of VGG on the test set: {accuracy_vgg:.2f}%')
print(f'Average loss of VGG on the test set: {average_test_loss_vgg:.4f}')


Using device: cuda
Files already downloaded and verified
Files already downloaded and verified
Training finished
Accuracy of VGG on the test set: 80.19%
Average loss of VGG on the test set: 0.5711


# Analysis and Discussion

## Comparison of Test Set Accuracy and Loss:

- **MLP**:
  - Accuracy: 53.67%
  - Average Loss: 1.5095

- **CNN**:
  - Accuracy: 67.83%
  - Average Loss: 1.3539

- **VGG-based Model**:
  - Accuracy: 80.19%
  - Average Loss: 0.5711

## Reasons Behind Performance Differences:

### CNN vs. MLP:
CNNs perform better than MLPs in image classification tasks due to their ability to leverage the spatial structure of images. CNNs use convolutional layers to extract features hierarchically, preserving the spatial relationships between pixels. This allows CNNs to capture local patterns and spatial dependencies, leading to better feature extraction and classification performance compared to MLPs, which treat images as flattened vectors and do not consider spatial information.

### VGG-based Model vs. CNN:
The VGG-based model, which is a deeper CNN architecture with more complex feature extraction capabilities, outperforms the simpler CNN in terms of accuracy and loss. The VGG model has learned rich and diverse features from a large-scale dataset like ImageNet during pre-training. Transfer learning from the pre-trained VGG model provides a significant advantage by initializing the network with meaningful weights, allowing the model to converge faster and achieve better performance on the CIFAR-10 dataset.

## Benefits of Transfer Learning with VGG:

- Transfer learning with the VGG model has helped improve both accuracy and loss on the CIFAR-10 dataset. The pre-trained VGG model has already learned to recognize a wide variety of features from ImageNet, which are transferable to CIFAR-10.
- The VGG model's hierarchical feature extraction capabilities, combined with transfer learning, enable the model to learn more complex and discriminative features specific to CIFAR-10 classes, resulting in improved accuracy.
- Additionally, transfer learning reduces training time and computational resources required to train a deep model from scratch. The VGG model serves as a powerful feature extractor, and fine-tuning the model on CIFAR-10 allows it to adapt to the specific characteristics of the dataset while leveraging the general knowledge learned from ImageNet.

In conclusion, CNNs, especially deeper architectures like the VGG model with transfer learning, outperform MLPs in image classification tasks due to their ability to capture spatial features and leverage pre-trained knowledge from large-scale datasets. Transfer learning with pre-trained models like VGG can significantly improve performance, reduce training time, and enhance the overall effectiveness of deep learning models on specific tasks.
